crates/astria-gossipnet/src/network.rs

```diff
@@ -1,77 +1,84 @@
+/// gossipnet implements a basic gossip network using libp2p.
+/// It currently supports discovery via bootnodes, mDNS, and the kademlia DHT.
 use std::{
     collections::hash_map::DefaultHasher,
     hash::{
         Hash,
         Hasher,
     },
-    pin::Pin,
-    str::FromStr,
-    task::{
-        Context,
-        Poll,
-    },
     time::Duration,
 };
 
-/// gossipnet implements a basic gossip network using libp2p.
-/// It currently supports discovery via mdns and bootnodes, and eventually
-/// will support DHT discovery.
 use color_eyre::eyre::{
+    bail,
     eyre,
     Result,
     WrapErr,
 };
-use futures::StreamExt;
-pub use libp2p::gossipsub::Sha256Topic;
-#[cfg(feature = "mdns")]
-use libp2p::mdns;
 use libp2p::{
     core::upgrade::Version,
     gossipsub::{
         self,
-        Message,
         MessageId,
-        TopicHash,
     },
-    identity,
+    identify,
+    kad::{
+        record::store::MemoryStore,
+        Kademlia,
+        KademliaConfig,
+        QueryId,
+    },
+    mdns,
     noise,
     ping,
     swarm::{
+        behaviour::toggle::Toggle,
         NetworkBehaviour,
         Swarm,
         SwarmBuilder,
-        SwarmEvent,
     },
     tcp,
     yamux,
     Multiaddr,
     PeerId,
     Transport,
 };
-use tracing::{
-    debug,
-    info,
+pub use libp2p::{
+    gossipsub::Sha256Topic,
+    identity::Keypair,
 };
+use multiaddr::Protocol;
+use tracing::info;
+
+pub use crate::network_stream::Event;
+
+const GOSSIPNET_PROTOCOL_ID: &str = "gossipnet/0.1.0";
 
 #[derive(NetworkBehaviour)]
-struct GossipnetBehaviour {
-    ping: ping::Behaviour,
-    gossipsub: gossipsub::Behaviour,
-    #[cfg(feature = "mdns")]
-    mdns: mdns::tokio::Behaviour,
+pub(crate) struct GossipnetBehaviour {
+    pub(crate) ping: ping::Behaviour,
+    pub(crate) identify: identify::Behaviour,
+    pub(crate) gossipsub: gossipsub::Behaviour,
+    pub(crate) mdns: Toggle<mdns::tokio::Behaviour>,
+    pub(crate) kademlia: Toggle<Kademlia<MemoryStore>>, // TODO: use disk store
 }
 
 pub struct NetworkBuilder {
     bootnodes: Option<Vec<String>>,
     port: u16,
-    // TODO: load key file or keypair
+    keypair: Option<Keypair>,
+    with_mdns: bool,
+    with_kademlia: bool,
 }
 
 impl NetworkBuilder {
     pub fn new() -> Self {
         Self {
             bootnodes: None,
             port: 0, // random port
```

Comment:
nit (not for this PR): using a default port of `0` is probably too random.

Reply:
yes probably :p